### ISSUE TITLE
[android] The active track recorder 

### ISSUE THREAD

**Issue by rtsisyk** [organicmaps/organicmaps#2147]
This ticket is for tracking the "active" track recorder. It should have at least start/stop buttons and display recorded data in kml/bookmarks UI.

See https://github.com/organicmaps/organicmaps/issues/512#issuecomment-1007939433 and https://github.com/organicmaps/organicmaps/pull/1807#issuecomment-1045934270 for previous research in this area.

Ideally, tracks should be exported as GPX, and uploaded into OSM/

## Upvote & Fund

- We're using [Polar.sh](https://polar.sh/organicmaps) so you can upvote and help fund this issue.
- We receive the funding once the issue is completed & confirmed by you.
- Thank you in advance for helping prioritize & fund our backlog.

<a href="https://polar.sh/organicmaps/organicmaps/issues/2147">
<picture>
  <source media="(prefers-color-scheme: dark)" srcset="https://polar.sh/api/github/organicmaps/organicmaps/issues/2147/pledge.svg?darkmode=1">
  <img alt="Fund with Polar" src="https://polar.sh/api/github/organicmaps/organicmaps/issues/2147/pledge.svg">
</picture>
</a>

**Comment by endim8** [comment]
The possibility to upload the kml to [osm traces](https://www.openstreetmap.org/traces) directly from the app would be cool

**Comment by anant01tyagi** [comment]
Hello @rtsisyk . I am highly interested to work on this as part of GSOC '22. How do I get started and what should I do to strengthen my chances?

**Comment by biodranik** [comment]
@anant01tyagi You need to know C++ for that, as we want to reuse the same implementation also for iOS. Here is a starting list:
1. Check docs
2. Build the project and run unit tests
3. Read the code, try to fix some issues (ideally related to location), and create a PR
4. Discuss with us a plan for the track recorder implementation.

**Comment by huuhaa** [comment]
This feature would be cool indeed. I hope if/when implemented, it will give option to give name to track. Will make it easier to add streets to OSM when you can check streetname from sign and start recording track with name of street. Later just add street to OSM using computer,

**Comment by axet** [comment]
Why not take it from here:

* https://gitlab.com/axet/omim/-/commit/108facda0eaddfda8f8df7b4e79481ef056d0b30

**Comment by biodranik** [comment]
Because there are many restrictions from Android OS. Only a properly implemented service with foreground notifications will work.

**Comment by axet** [comment]
It is already has foreground notification. Have you ever looked at the code?

**Comment by biodranik** [comment]
Of course. The code base changed significantly in 5 years. Also, you used some permissions that we won't use. The idea is to implement something similar, it won't be possible to blindly apply the patch.
CC @rtsisyk

**Comment by shankarpriyank** [comment]
I had a look at #512 and #1807.
And did some digging around location tracking and https://dontkillmyapp.com/

Summing it up

- Using BACKGROUND_LOCATION permission is not a good idea, there are many reasons for that (google won't approve it, it's best for the user to actively know that their location is being tracked, and with background tracking reliability will be a big issue)
-  The best way to implement a tracking feature is by using a Foreground Service, we won't need extra permissions for it too, so eventually integrating this feature in the app will also not create a complex permission handling situation. One challenge that I can see with a tracking feature is that if we allow the user to pause their tracking this may lead to useless/meaningless data so maybe we will have to do some post-processing of the data captured.
- I made a sample/POC app to test the tracking out and it works just fine on my device. It uses a foreground service with an ongoing notification.   https://github.com/shankarpriyank/BackgroundLocationTracking
- Though even with foreground services 100% reliability is not guaranteed, but I don't this should bother us, cause while tracking the location we are also expecting the user to open their device once in a while for routing purposes and that with the use of foreground service could work well. Power/Battery Saver mode can mess with the functionality, Maybe we can use a broadcast receiver to gracefully stop tracking in that scenario. While testing my sample app the foreground service was alive for more than 7hr without my touching the device.
- About the killing of foreground services, and I don't think there is any way to guarantee 100% reliability, if we want 100% reliability, we will have to do A/B testing. There are some generic hacks to follow though for best results across devices, like making your service sticky and prompting the user to whitelist the app from battery optimizations. But still, no guarantee that the service will live.
- Also I guess to implement this feature we will need some UI changes, I don't have any ideas rn, where would it be the best to add this feature of tracking the location. Would love to hear different views on this.

Also, I can't really see the need for C++ code here, I am pretty confident that even the conversation of the data captured into a meaningful format like GPX can be done using Java/Kotlin, performance can be an issue though.

Please let me know if I am missing something @biodranik

**Comment by biodranik** [comment]
1. Please test/take into account the following scenarios while testing: the battery saving option on the device is set to the most aggressive one + your device's battery goes low (e.g. <20% or <10%).
2. Please think of and design a good UI/UX for the foreground service. It is _a must_ before starting to implement it. It should be easy to use and allow users to keep the current (no background) behavior if necessary (I'm speaking not only about the Track Recorder here, but also about the background navigation directions too because they are closely related: in both cases location should be updated while the device is locked).
3. C++ part (with unit tests) to process and store recorded data is absolutely necessary, to avoid code duplication on iOS and other platforms.

**Comment by shankarpriyank** [comment]
1. Okay, will test the app in the mentioned scenarios
2. UI/UX for track recorder is one thing, but if we use a foreground service for navigation too, we will be able to track the user even when the device is locked or when the app is in the background, and with all of this it will be possible to deliver TBT notifications and it the user can navigate easily even without needing to open the app(gmap uses TBT very well). But that is a completely different thing, and will require a lot of work and meticulous planning. I think implementing a track recorder will require significantly less efforts than implementing a proper TBT navigation system using foreground service. 
3. Got it, but this [comment](https://github.com/organicmaps/organicmaps/issues/512#issue-912839317) says that this feature(track recorder ) is already available on this iOS app

**Comment by biodranik** [comment]
\2: We can start without text notifications (although they are easily implemented on top of our existing TTS voice strings), just by using the voice while in the background. That feature is requested by many users, hikers, drivers, and cyclists.
\3: iOS has a "recent track" feature which is unstable, and doesn't allow now to save/export of the recent track that disappears in a day.

**Comment by shankarpriyank** [comment]
Okay, I get it, using a foreground service is a necessity and we can use the ongoing notification for showing the user information related to routing. I think we should focus more on the implementation strategy, cause that looks challenging to me, and as far as UI/UX is concerned I don't think we are going to make any changes to the navigation ux, it will be just that the user will able to see navigation info on the notification bar too(we can take inspiration from gmap for it).
And for the track recorder part, where do you think will it be best to put the option to start/pause/stop tracking?( not focusing on where can the user find the previously recorded tracks rn).

Also one small thing regarding UI, what exactly should we show in the notification while the app is tracking user's location?( for the track recorder). While navigating it makes sense to show the navigation info, and I think it's best to show options to pause/stop the tracking in the notification itself for the track recorder.

And I think it's best if we could get on a call to discuss some implementation strategies/ideas cause as far as I can see it is going to be challenging.  And we could also discuss the [plans](https://github.com/organicmaps/organicmaps/discussions/4342#discussioncomment-5021319) for Android Auto

**Comment by shankarpriyank** [comment]
I did some testing too, on my device POCO X2 there is a battery saver mode and there is ultra battery saver mode.
I tested the sample app while my device was on battery saver mode and it worked fine, the service was running  for more than  4hr.
As far as ultra battery saver mode is concerned, it is very aggressive it will kill any active services running. And even the user interaction with the device is heavily restricted.
Overall I would say tracking is working fine in all scenarios on my device

**Comment by biodranik** [comment]
Is it possible to disable power saver mode for a specific application? If yes, then from which Android version? Are there any manuals about it? How other track recording apps are hinting users about it?

Let's prepare a UX idea of what we can suggest to our users to avoid any issues and surprises when using the Track Recorder feature. Where and when these hints should be shown? Do we need a more detailed web page with instructions and screenshots for different devices/manufacturers?

**Comment by shankarpriyank** [comment]
Regarding the Android version I will have to do some research. 
Could not find any manuals but these 2 threads seems to give a nice overview of the approach

- https://stackoverflow.com/questions/41722516/disabling-or-detecting-the-app-specific-power-saving-mode
- https://stackoverflow.com/questions/48166206/how-to-start-power-manager-of-all-android-manufactures-to-enable-background-and/48641229#48641229

UX is a very subjective thing, I guess it will need some discussion.

**Comment by biodranik** [comment]
About the UX changes for the track recorder:
1. Current bookmarks button icon can be modified to show that users can also find recorded and imported tracks there.
2. Start track recording can be placed differently:
   1. On the main screen (but not everyone may need it there)
   2. In the menu (more clicks)
   3. It can be shown when users select a current position arrow (hard to discover)
3. Start track recording should be also easily accessible when active navigation mode has started. Maybe we can even add a setting "always record a track while navigating".
4. Start button can be converted to Pause button.
5. What about the Stop and Resume buttons? How other apps are doing it conveniently? Please check Strava, OSMand, [GPS logger](https://play.google.com/store/apps/details?id=eu.basicairdata.graziano.gpslogger) and other popular apps.
6. When tracks should be split into different segments? Should they go into the same bookmarks list? (likely yes, but we need to implement moving tracks between lists).
7. What will be the name of the recorded track?
8. Foreground Service notification should have pause/start/stop buttons, definitely. Maybe some useful info can also be displayed (e.g. the length of the recorded track).

**Comment by sum20156** [comment]
<img width="974" alt="image" src="https://user-images.githubusercontent.com/35376901/224462684-4672494b-80e2-4718-a8a8-77d3f6c95832.png">

- we need `ACCESS_BACKGROUND_LOCATION ` permission even we use foreground service.
- Another thing want to mention **Apps that target Android 12 (API level 31) or higher can't start foreground services while running in the background** except few situation like if battery optimisation is disabled. Are we having any use case in app where we may need to start the Foreground service from background?

**Comment by shankarpriyank** [comment]
I don't think there is such a use case. The user will always start the tracking manually which implies that the app will be in foreground that time. So we won't need that permission

**Comment by sum20156** [comment]
@shankarpriyank I think there is miscommunication , two point I mentioned is not co-related. They are standalone point.

**Comment by shankarpriyank** [comment]
The image which you attached says that the restrictions on the foreground services are imposed only if it's triggered while the app is in background.
And I don't think we would need to trigger a foreground service while the app is in background, we will trigger it while the app is in foreground. That is why we won't need that permission.

Sorry if I was not clear in the last comment, I hope we are on the same page now @sum20156

**Comment by sum20156** [comment]
yes @shankarpriyank

**Comment by sum20156** [comment]
![WhatsApp Image 2023-03-11 at 10 06 54 AM (1)](https://user-images.githubusercontent.com/35376901/224466190-f31ec9ab-8244-4459-9f43-e1e44ed2ff3c.jpeg)

![WhatsApp Image 2023-03-11 at 10 06 54 AM](https://user-images.githubusercontent.com/35376901/224466412-285a1514-cad2-4f91-aa03-4665c9e556e1.jpeg)

Hi @biodranik GPS logger, Strava doesn't have a pause button. OSMAnd has a start/stop button and pause button with different UI for pause, and also quite useful information about the trip in the notification.

**Comment by biodranik** [comment]
@sum20156 Thanks! Let's research the convenience/necessity of the "Pause" button.
1. Why is it there in OSMand? Why other trackers don't use it?
2. Which real-world issues does it solve? Why do users find it useful?
3. How does it behave? E.g. does it record into the same track or create a new one? Does that recorded track have gaps in it?

Let's focus on the best possible UX in Organic Maps that will be more convenient than in other apps )

**Comment by sum20156** [comment]
### **1. Why is it there in OSMand? Why other trackers don't use it?**
When using Strava, clicking the "stop" button doesn't discard the tracking session. Instead, it pauses the session, and clicking "start" resumes it. However, you can only finish the session from the app's screen, not from a notification.

On the other hand, with OSMand, clicking "stop" will discard the tracking session, which is why they have a separate "pause" button.

### **2. Which real-world issues does it solve? Why do users find it useful?**
A pause feature can be helpful in a GPS track recorder app for several reasons:

1. To conserve battery life: When the user pauses the tracking, the app stops using the GPS and other sensors, which can help save battery life. This is especially useful for activities that take a long time, such as hiking or marathon running.

2. To exclude breaks or stops: If the user takes a break or stops during their activity, they can use the pause feature to exclude that time from the recorded activity. This can give the user a more accurate record of their performance, especially for activities where breaks or stops are common, such as cycling.

3. To avoid clutter in the data: If the user accidentally leaves the tracking on while they are not actually moving, the recorded data can become cluttered with unnecessary information. By using the pause feature, the user can avoid this and keep their recorded data clean and accurate.

4. To avoid errors in tracking: If the user goes through areas where GPS signals are weak or unavailable, the tracking may become inaccurate or even stop altogether. The pause feature allows the user to stop tracking during these times to avoid errors in tracking.

### **3. How does it behave?**
I have checked Strava and MapMyWalk app, and they do create a gap when we hit the pause button, but when we resume, it fills the gap.

### Some good ideas for making it very convenient for the user

1. Shortcuts: Consider adding hardware key shortcuts to make it easy for users to start, stop, and pause tracking. For example, assign one of the volume keys to start, stop, or pause tracking. This can be done by allowing users to customize the function of the volume keys in the app settings.

2. Voice commands: Voice commands can make the app more convenient for users who are in the middle of an activity and have their hands full. For example, users could say "start tracking" or "pause tracking" to control the app.

3. Auto Start/ Auto Pause: Consider adding an auto-start feature that automatically starts tracking when the user begins an activity. This will eliminate the need for the user to manually start tracking each time. An auto-pause feature can be helpful for activities where the user may stop or take a break, such as cycling or running. This feature automatically pauses the tracking when the user stops moving and resumes tracking when they start moving again.

**Comment by shankarpriyank** [comment]
Nice ideas. One thing I want to point out is that adding voice commands will be a completely different vertical. To provide a smooth UX we will need to capture the user's voice while the app is background (implementing this would be a challenge). 
Also personally I don't think it's a good idea, cause even if we are listening to the user in background I don't this its realistic to expect that the mic will be in range to capture the user's voice. Expecting the user to open the app and then use voice commands doesn't make sense. To me clicking a button is easier than saying a few words. But these are just my personal opinions.

**Comment by biodranik** [comment]
@sum20156 thanks for the thorough answers! I like strava's approach, without a Pause button, but with a Stop button that behaves like a Pause. It is always easy to extend and overload functionality, while it's hard to keep it as simple and as useful as possible.

**Comment by sum20156** [comment]
hi @biodranik here are my perspective regarding **UX/Ui** for this feature:

### Where to place start button?
1. Consider using a menu or navigation drawer to provide access to the start recording button, as it is not a primary action. This can help keep the user interface clean and uncluttered.

2. Place the start recording button in a location that is consistent with the design of other apps in the same category. Users may expect to find certain buttons or navigation elements in specific locations based on their prior experience with similar apps.

### What will be name of the track?

**Pros of Asking for a Name:**
 
- Helps users identify and organize their tracks more easily.

- Allows users to add a personal touch to their tracks and make them more meaningful.

- Can make it easier for users to search for and find specific tracks later on.

**Cons of Asking for a Name:**

- May add unnecessary complexity to the tracking process, especially if users are just using the app for a quick, one-time tracking session.

- Users may forget to name their tracks or may not want to spend the time thinking of a name.

- If users don't name their tracks,  will need to come up with a default naming convention based on the date, time, or location of the recording (e.g., "Track  Central Park 2022-05-03 12:34:56")  to avoid having multiple tracks with the same name.

> I will suggest not to ask for a name but provide the option to edit it later as tracking is not core functionality .


### Information we can show in notification

- Current location: Displaying the current location of the user can help them stay oriented and avoid getting lost.

- Distance to destination: Displaying the distance remaining to the destination can help users track their progress and estimate how much time they have left.

- Estimated time of arrival (ETA): Displaying the estimated time of arrival can help users plan their route and make any necessary adjustments.

- Turn-by-turn directions: Providing turn-by-turn directions, including the next upcoming turn or direction in the notification can be helpful for users.

- Current distance: Displaying the current distance traveled by the user can be a useful way to keep them updated on their progress.

- Elapsed time: Displaying the elapsed time since the user started tracking can be helpful for users who are trying to keep track of their workout or exercise routine.

- Average speed: Displaying the average speed of the user can help them monitor their progress and make adjustments to their pace if necessary.

> We can provide option in setting to customise notification.

**Comment by sum20156** [comment]
Hi @biodranik I'm interested in contributing to the development of the this feature in the app. Would you be able to suggest any specific issues or areas of the codebase that I could explore in order to gain a deeper understanding of how navigation is currently implemented in the app? I appreciate your assistance and look forward to the opportunity to contribute to the project. Thank you.

**Comment by biodranik** [comment]
@sum20156 Thank you. It would be great to be _more specific_ on your ideas and suggestions, and propose specific locations for buttons/labels in the existing UI. For that, you need to use OM and other apps actively and should understand/feel what would be _easier_ and _more convenient_ to users.

1. Keep in mind that Track Recorder should be implemented on Android and iOS (it can be done separately though).
2. There are several possible locations for the Start/Stop recording button now. To highlight this feature to as many users as possible, it can be even placed on the main screen. Also don't forget that the same button should conveniently _stop_ recording (or pause it, as it was discussed above).
3. Asking for the track name is not the best UX.
4. In addition to the timestamp, you proposed to insert some autogenerated text into the track name. How you will get this text? What should be the algorithm to extract this text?
5. Track recording takes some time. Should a time range be displayed in its autogenerated name, instead of the start date?
6. Should the track name (autogenerated date/time range) be localized and made easier to read than in ISO Timestamp format?
7. Can you please elaborate on "Show current location in notification"? Users already see where they are by looking around. What do you want to display in the notification? Coordinates? Why do you think that it is helpful?
8. What kind of destination, ETA, and directions are you talking about? What if the user started track recording _without_ building any specific route and without navigating to the destination? Looks like you mixed up the track recording with the active navigation notification. For an appropriate implementation, we should design and support the following scenarios for the notification:
   1. Only navigation is active, without a track recorder (your ideas likely apply here).
   2. Only the track recorder is active, without navigation (the UI/UX should be different, right?)
   3. Both navigation mode _and_ track recorder are active. The complex case, where it would be great to still use one notification panel, if it is possible, of course. Or we can split and support two independent notifications (will it be convenient for users?)

**Comment by biodranik** [comment]
There is no background navigation yet in OM, so the notification should be designed too.

The most complex part of the task for Android is to design the lightweight background location service, considering https://dontkillmyapp.com/

**Comment by shankarpriyank** [comment]
@biodranik To show both the info of the track recorder and that of navigation in a single notification panel may end up looking cluttered. I vote for 2 different notification panels for two reasons.

- Easier Implementation
- The track recorder notification panel will have to show its own information like the time for which the tracker has been active and the distance that has been tracked and a button for stopping/pausing the tracking. And then there is the navigation info that the navigation notification is supposed to show, it's a lot in itself. Combining them together may create a clutter making it hard for the user to comprehend the information.

For reference purpose, this is how the notification panel for navigation of gmaps look like.  Is there space to show more info here?
Different buttons with completely different objectives may also be confusing on a single notification panel
![image](https://user-images.githubusercontent.com/75121767/225090950-7056a34e-3329-4477-8bd3-b52418c8f03d.png)

**Comment by sum20156** [comment]
### I prepared a doc by observing all popular applications which has track recorder feature. You can take a look here: 
[track recording examples.pdf](https://github.com/organicmaps/organicmaps/files/10986529/track.recording.examples.pdf)

### Where to place start button?
Major apps have placed the start button at the bottom, so to maintain familiarity we can place start/stop button at the bottom along with other buttons. We can place a view with white bg at the bottom to show the recording details like avg speed etc, it will be similar to what we show while navigation.

![image](https://user-images.githubusercontent.com/35376901/225514617-cd485dbc-db3b-4a98-b777-1b13866fc583.png)

### What will be the track name?
I think we can do similar to what [Cyclers](https://play.google.com/store/apps/details?id=com.umotional.bikeapp&hl=en) is doing. Check page no 7 of my attached pdf. (Ride to {current location}).  Including location information in the track name can help users quickly identify where the track was recorded. This can be especially helpful for users who record tracks in different locations and want to keep them organized and it can make the track recording experience feel more personal and customized to the user's location. It can also help users remember the places they've visited and the paths they've taken. Will it be possible to localize?

### Regarding notification:

- Cluttered notification area: Having two separate notifications can clutter the user's notification area, especially if they are also receiving other notifications from other apps. This can make it harder for users to find the information they need and could lead to frustration or confusion.

- User confusion: If the notifications are not clearly labeled or separated, users may become confused about which notification corresponds to which function. This could lead to errors or misunderstandings, such as accidentally stopping navigation instead of track recording.

We have to show minimal information on notifications, we can discuss what are those and also provide users to enable/disable them from settings based on his/her preferences.

**Comment by biodranik** [comment]
Thanks for the feedback.

1. It is not fair to compare recording in OM (a general map app) with recording in narrow-focused apps for cycling or running only. A better comparison would be with other general map applications.
2. Track recording is not the _main_ feature of OM, like in some other apps. It is a _supplementary_ feature, that may not be used by the majority of users. From the other hand, it would be great to tell users that this feature exists by clearly indicating it in the UI.
3. Should any other information about the active track/recording status be displayed on the main map screen? Or only a record button that changes to a stop button (when recording is active) is enough? Will the foreground service notification be visible anyway when OM is actively recording? Can we show all recording-related information only in the notification, to avoid the main screen pollution?
4. What exactly should happen when the stop button is pressed? Should we show any UI? Won't it be too intrusive? Or maybe allow optional track naming somehow?
5. It is not always clear where the user started or finished track recording (it may happen on the empty map). How tracks should be named in this situation?
6. If we insert names/addresses of nearby objects into the recorded track name, how to choose the best object if there are many of them? Or should it be the address only? What if there is a street, without a house number?
7. Should POIs/addresses near the starting point be added? Or only near the ending point?
8. How recorded tracks should be organized? Should they all be added to some "default" list? Or should each track go into a different list? Or should different lists be used depending on some other criteria (transport type, avg. speed, place where the track was recorded, etc.)? How do other apps help to simplify track management and storage? What about Strava? OSMand? [GPS Logger](https://play.google.com/store/apps/details?id=eu.basicairdata.graziano.gpslogger) ?
9. It would be great to see a prototype with notifications for navigation and a track recorder both active in OM, to understand how convenient is it.

**Comment by shankarpriyank** [comment]
- My guess is that we can add all the needed information in the notification itself.
- Instead of naming the file based on address or objects, we can name them using the date and time, that would make sense too.
- I will try to prepare a prototype(Making a prototype would be challenging as background navigation is also not implemented) and maybe we can use dummy notifications just to under about the UX

Regarding features like track management and allowing optional naming, I think it's wise to not focus on them now and plan to add them over in the next iterations

**Comment by biodranik** [comment]
Does foreground service require granted notification permission from a user for track recording and navigation? If yes, then please propose an appropriate and convenient UX for users to enable if it wasn't enabled, or if it was disabled for some reason.

**Comment by sum20156** [comment]
Hi @biodranik, here is the prototype with notifications for navigation and a track recorder both active in OM.

https://user-images.githubusercontent.com/35376901/227104101-2f1d01d7-6af6-4f83-8a4c-00f329bfe655.mp4

**Comment by biodranik** [comment]
@sum20156 thanks! Notifications should be carefully designed to look different (if we use two of them) and be convenient to use.
What does OSMand show for track recording? What does it show for both navigation + track recording active? Are there any good notification UX examples from other apps, except those you listed?

**Comment by pranjalsrv123** [comment]
@biodranik please check out this one, it's executed in android studio.
![maps1](https://user-images.githubusercontent.com/103902468/227739602-eb178c86-3a06-46c8-ad81-d33ebacd2fe5.jpg)

![maps11](https://user-images.githubusercontent.com/103902468/227739613-b32e91d9-a5a6-42f3-a766-30712bfe52b9.jpg)

**Comment by pranjalsrv123** [comment]
@biodranik this one also please. All codes are written and executed in android studio.
![map21](https://user-images.githubusercontent.com/103902468/227739844-ba627f50-4b9c-4173-9035-e71d122b7cb6.jpg)
![mp23](https://user-images.githubusercontent.com/103902468/227739864-ca0b92f0-bf6e-4151-9ee1-e9e16ad99776.jpg)

**Comment by sum20156** [comment]
Hi @biodranik here are the apps I found where navigation and recording can be used together or individually. 

### [OsmAnd](https://play.google.com/store/apps/details?id=net.osmand&hl=en&gl=US)

**Navigation + Track Recording**
![image](https://user-images.githubusercontent.com/35376901/227751984-ce85e209-fd65-40a0-bb04-a0f2dd4f383d.png)


### [Mapy.cz](https://play.google.com/store/apps/details?id=cz.seznam.mapy&hl=en&gl=US)

**Navigation + Track Recording**

![image](https://user-images.githubusercontent.com/35376901/227753238-9ea893c6-a9a5-4700-9377-b6b34898ba8f.png)

### [Locus Map](https://play.google.com/store/apps/details?id=menion.android.locus&hl=en&gl=US)

**Navigation + Track Recording**
![image](https://user-images.githubusercontent.com/35376901/227754844-5b3aec27-d606-4514-9bfe-dbc14673452c.png)

### [Guru Maps](https://play.google.com/store/apps/details?id=com.bodunov.galileo&hl=en&gl=US)

**Navigation + Track Recording**
![image](https://user-images.githubusercontent.com/35376901/227754001-3207c987-e277-4dcb-9a6a-2a4126c1979d.png)

Most of the apps are either only navigation-based or only track recording based. I found very few apps that are capable of both at a time. If you know any special keywords for searching these types of apps let me know.

- Regarding notification permission on android 13, Without POST_NOTIFICATIONS permission foreground service can be started & running but notification will not be shown.

**Comment by biodranik** [comment]
@sum20156 thank you for the screenshots! Are there any buttons if you expand recording notifications in Locus and Mapy?

**Comment by sum20156** [comment]
> @sum20156 thank you for the screenshots! Are there any buttons if you expand recording notifications in Locus and Mapy?

yes for locus it has a pause and a stop button and in Mapy it has only a pause button(On pause click notification disappear).

**Comment by shankarpriyank** [comment]
> Hi @biodranik here are the apps I found where navigation and recording can be used together or individually.
> 
> ### [OsmAnd](https://play.google.com/store/apps/details?id=net.osmand&hl=en&gl=US)
> **Navigation + Track Recording** ![image](https://user-images.githubusercontent.com/35376901/227751984-ce85e209-fd65-40a0-bb04-a0f2dd4f383d.png)
> 
> ### [Mapy.cz](https://play.google.com/store/apps/details?id=cz.seznam.mapy&hl=en&gl=US)
> **Navigation + Track Recording**
> 
> ![image](https://user-images.githubusercontent.com/35376901/227753238-9ea893c6-a9a5-4700-9377-b6b34898ba8f.png)
> 
> ### [Locus Map](https://play.google.com/store/apps/details?id=menion.android.locus&hl=en&gl=US)
> **Navigation + Track Recording** ![image](https://user-images.githubusercontent.com/35376901/227754844-5b3aec27-d606-4514-9bfe-dbc14673452c.png)
> 
> ### [Guru Maps](https://play.google.com/store/apps/details?id=com.bodunov.galileo&hl=en&gl=US)
> **Navigation + Track Recording** ![image](https://user-images.githubusercontent.com/35376901/227754001-3207c987-e277-4dcb-9a6a-2a4126c1979d.png)
> 
> Most of the apps are either only navigation-based or only track recording based. I found very few apps that are capable of both at a time. If you know any special keywords for searching these types of apps let me know.
> 
> * Regarding notification permission on android 13, Without POST_NOTIFICATIONS permission foreground service can be started & running but notification will not be shown.

The notification OsmAnd seems interesting, save is equivalent to stop I guess?

**Comment by sum20156** [comment]
yes @shankarpriyank

**Comment by biodranik** [comment]
Great! Let's document now the UX on what happens if the user pauses/stops/saves track in the apps above. In which cases which notification disappear? How the same functionality is done in each app?

Another question: There is a "recent track" feature now in iOS (it was disabled on Android). It _always_ records user movements until user explicitly disables it and draws a recent (now up to 24H) track on the map.
It is now done without great precision to save some battery life (but looks like it may be useful to allow configuring it in both "battery-saving" and "max precision" modes). Will it have a _separate_ notification (a third one)? Won't it be too much? Are there any better ideas to have no more than two notifications?

**Comment by shankarpriyank** [comment]
Maybe I did not get it, but why do we need a separate notification, if we have two option, we can add an option to choose between them when the user starts the tracking

**Comment by biodranik** [comment]
If a user explicitly enabled "recent track", and then enables (precise) "track recording", asking him to choose only one of two would not be a good UX. Many users have already gotten used to the "Recent track" functionality, and it is an easy way to not get lost in an unknown place.

**Comment by sum20156** [comment]
I checked the IOS app, and here is what I understood:
so basically, it records the track in the background and draws it on the map, and after a certain interval, it gets deleted. In IOS we don't need to show any notification, but in android, due to background restrictions, we have to use notification. So we can use the same notification id for recent track and track recorder foreground service. Basically, we will have two states for that notification. For recent tracks will just show simple text like "Recording location for recent tracks" and when the user will start recording, will make recording information visible along with that text. While updating the notification we have to add a check isTrackRecorderRunning() and based on that state will be set.

**Comment by pranjalsrv123** [comment]
@rtsisyk please this one. It works offline and gives correct navigation.
![Screenshot (16)](https://user-images.githubusercontent.com/103902468/228146340-806ffe64-f8d1-464a-9698-5d7150edc4b6.png)

**Comment by Aurareus** [comment]
I'm currently using the app Trekarta for this functionality but I'm excited to be able to fully replace it with Organic Maps once this feature is implemented.

Here is how the UI/UX implementation is in Trekarta (I'm certainly not suggesting this as the optimal solution, it's very hidden/difficult to figure out as a new user, but perhaps it could be useful just as another example of how it's been done in other apps):

![Screenshot_20230530-010353_Trekarta](https://github.com/organicmaps/organicmaps/assets/19821158/5d7b505b-7759-4cd7-a064-bf6ea4f322eb)
The navigation bar at the bottom has a button dedicated to recorded trails, when selected it shows a list of all recorded trails as shown in this screenshot which allows you to toggle visibility of the trail path on the map (eye icon on the right of each trail).
Selecting any individual trail brings up a details page for that trail:

![Screenshot_20230530-010622_Trekarta](https://github.com/organicmaps/organicmaps/assets/19821158/3e53be9f-6584-442f-8014-f5f54b2ef054)

In order to start recording a new trail, you have to long-press the trail button, which makes the trail button lit up/highlighted with an accent color:

![Screenshot_20230530-010413_Trekarta](https://github.com/organicmaps/organicmaps/assets/19821158/527ca7f8-df4f-4002-8178-68d3251f1b13)
This is what it looks like in the app when a trail is currently being recorded.
(Screenshot at random location without map data for privacy reasons.)

Here is what the notification looks like:

![Screenshot_20230530-010448_Trekarta_1](https://github.com/organicmaps/organicmaps/assets/19821158/b869716e-b6b4-4069-8065-2b0d3a64ced0)

And here is what the notification looks like when expanded, with options to pause or stop the trail recording:

![Screenshot_20230530-010501_Trekarta_1](https://github.com/organicmaps/organicmaps/assets/19821158/bcbbc5f3-4f33-4a09-a581-65d63c5bd814)

Stopping a recording from within the app UI is done the same way as starting it (long-pressing the trail button), which brings up a details page for the currently in-progress trail:

![Screenshot_20230530-011210_Trekarta](https://github.com/organicmaps/organicmaps/assets/19821158/023c86f7-b05a-47c3-b431-577847194eef)
The option to stop recording/finish the trail is the icon in the top right, next to the three dot icon.

Another thing that might be worth mentioning is that I personally find the ability to pause a trail very helpful, my use case for it is that I like to record only the trails I walk.
Often I will walk for a bit and then pause the tracking as I get on a bus or a train and then unpause it once I get off so that only the parts of my journey that I actually walked are recorded as trails.

**Comment by Aurareus** [comment]
Additionally here is how the notifications look in Trekarta with both navigation and tracking active at the same time (uses two separate notifications, like other apps):

![Screenshot_20230531-134450_Fennec_1](https://github.com/organicmaps/organicmaps/assets/19821158/34c4b18d-aa88-4449-8c01-2b14120a2c4a)
![Screenshot_20230531-134519_Fennec_1](https://github.com/organicmaps/organicmaps/assets/19821158/decb8f19-8abd-4203-b2be-ec688ba06da8)


Some thoughts on implementation:
I think having both Pause and Stop as options on the notification is probably preferable, two options on the notification rather than one doesn't constitute clutter imo, but on the other hand I also cannot personally see a case where I would want to stop the track recording without opening the app, other users may find that a useful option though.
Perhaps this is a little pedantic but I think if you were going to include a Stop option on the notification it would communicate the function better/more unambiguously if it was labelled "Finish" instead of "Stop".

For track naming I think the default naming scheme used by Trekarta (date and time of start and end of track, see screenshot in previous post) with the optional ability for users to rename tracks is probably the optimal solution.
To me at least it would seem that the most useful information to have in the name of a track is when it was recorded.
Whereas using location information seems like it would be confusing and unhelpful if you regularly record tracks in the same locations (which I would assume to be common in how users would use the feature?), plus I think it could be difficult deciding how to algorithmically choose which POIs/location data to use for a track name that is reliably intuitive.
And GPS coordinates are just ugly, useless meaningless numbers to the majority of users.

Perhaps you could include a toggle in the settings for prompting for track name.
When the setting is enabled the user would be prompted to name a track immediately upon finishing the track (perhaps with the text box pre-filled with the default date and time naming scheme?), and when the setting is disabled the track would be saved with the default name without prompting the user (they could still manually rename the track later).

**Comment by DeathfireD** [comment]
I'm not sure if it's been mentioned but OpenTracks, an open source project strictly for recording activities, offers an API that you can tap into for starting, stopping, and pulling back data while it's recording. The developer also offers an open source companion app called OSM Dashboard that's essentially a less user friendly Organic Maps. OSM Dashboard uses the API to pull data from OpenTracks every few seconds (I think) to show "live updates" as you're doing an activity.

As of right now I've been using OpenTracks to record my activities and then using the "open with" option to open the recoding in Organic Maps to view it. Doing this will not reload the data every few seconds in Organic Maps. It only displays the data from the point that you chose to open it in Organic maps. It's essentially importing a KML/GPX file into Organic Maps. With that said, I think while you're fleshing out your own activity tracker for Organic Maps, it could be a good temporary solution to tap into OpenTracks API  so that Organic Maps can pull recording data in real time from it. Just like how OSM Dashboard does.

https://github.com/OpenTracksApp/OpenTracks
https://github.com/OpenTracksApp/OSMDashboard

**Comment by Aurareus** [comment]
I think cloning all the OpenTracks functionality within Organic Maps is probably the ideal implementation of this feature.

**Comment by J4CKED** [comment]
I've tried OpenTracks, but I never know when the GPS is working. 

I currently use maps.cz (cz.seznam.mapy) as a tracker. It is a very simple tracker, but it works.
The green button activates tracking. To save tracks you need to create an account, but I don't use this feature. I just take screen shots.
![Mapy cz01](https://github.com/organicmaps/organicmaps/assets/60573158/109f8799-02c8-4e81-b1e5-1e18cb6b49c0)
![Mapy cz02](https://github.com/organicmaps/organicmaps/assets/60573158/18d80690-eb1f-41c9-a6a9-cbb0436f0572)
![Mapy cz](https://github.com/organicmaps/organicmaps/assets/60573158/d74cf3ee-36df-4b02-996c-a06792dc651a)

**Comment by dandud100** [comment]
What about ["Routes" for Android](https://routes-navigation.de/) i used it to record all of my tracks, the UI got bloated more and more over time but that's a different issue. But I really liked it, as it worked really good, you should not forget that we need to implement a slider to determine how often the position gets saved to the file so that 90 degree curves not look like this.

(Routes didn't had this option at first, see what happened in my first recorded routes, meanwhile it it implemented.)

![Screenshot_20230725-201349](https://github.com/organicmaps/organicmaps/assets/43549788/f84b242a-3da9-4b17-b8a1-81ef3116b787)

**Comment by trvrcr** [comment]
On Nextcloud there is an app that is used to record all sessions.
It already implements with many phone apps.
It would be great if OM is added to the list.
I'm available to test iOS app if needed.

**Comment by OhBoy1** [comment]
This has been so long in the making. I wonder if it could be split into two. One - Navigation recording. Many times I deviate from the indicated route for various reasons and a recording would be of immense help in adding POI's & tracks later in OSM.
Two - Track recording as discussed above.
Intuitive I find option one simpler to implement as it is in foreground, GPS positions are known so all it needs is to be written to a GPX file.
I guess I  am too simplistic here but is just my thoughts.

**Comment by sandy99405** [comment]
Hi @biodranik  I'm very much interested in contributing to this as part of **Google Summer Of Code 2024**.

I've gone through all the issues and references related to the Track recorder and summed up the points to execute this feature.
Here are the steps: 

1) As the Track recording should only happen after enabling navigation, the start recording button though will be placed on the regular map_buttons_layout, it will be enabled only when the navigation starts.
(***Since we want the users to know that there is a track recording feature, the button will be placed on the regular layout and if pressed on it, a message will appear saying "Start Navigation to start recording"***)
**Regular :-**
![Start recording Debug Organic Maps](https://github.com/organicmaps/organicmaps/assets/55664068/5d42e519-90eb-471f-aa3d-d5a37794fd93)
**Navigation :-**
![Screenshot_20240317_081650_Debug Organic Maps](https://github.com/organicmaps/organicmaps/assets/55664068/159e29fd-7d21-435b-95f5-bc675eda893c)


2) The recording starts with the foreground service through a notification when the app is in the background.

3) The Notification bar contains a pause button (which converts to a resume) and a stop button(Note: when the pause button is clicked and the driver covers some distance, the recorder will fill the track based on the point where the user clicks on resume and continues with the user till the destination point). It also displays the distance covered and the distance yet to be covered.

4) Once the trip has been completed, the user stops the recording(Either from the notification or from the application) a dialog box appears just like the bookmarks list asking for a track name already filled with a name(date_time_start_destination). If the user wants to change the name they can edit the dialog box, or else they can save it as it is.
**I've used this Algorithm to generate record name**
![image](https://github.com/organicmaps/organicmaps/assets/55664068/ea96b9fe-7215-4bd4-9bd6-137f7619b1bc)

**Dilogbox example from Organicmaps Bookmarks List**
![Track name Debug Organic Maps](https://github.com/organicmaps/organicmaps/assets/55664068/6ec036d5-1ff3-4e8a-9027-0a5d3244ddb8)

5) After the user saves the track, the track will be available on the bookmarks list page. 
    For this, I have modified the bookmarks icon to this :
     
![image](https://github.com/organicmaps/organicmaps/assets/55664068/5a055e20-e7a2-4059-8a1a-841258c436b1)


6) Here I recommend a feature that I observed in a **non-open-source application "Komoot"**, which might impress you. 
     **1) When the track is saved there will be a template for each record.
      2) The user can later edit the record with highlights of the tour, add photos change the name of the record, etc. within this template.** 
![Screenshot_20240313_223634_Komoot](https://github.com/organicmaps/organicmaps/assets/55664068/66ba5353-5925-4763-8123-87bbc5d1e0bf)


Please let me know what you think about this.

**Comment by biodranik** [comment]
Track recording can be started at any time, independently from the navigation.

Automatically recording the track during the navigation is a useful feature that also can be implemented on top of the same functionality.

**Comment by sandy99405** [comment]
yes @biodranik , I agree with you on that.

1) When a user clicks on the record button(present on the regular screen) and starts a ride without starting navigation, the track will be recorded.

2) Let's add a switch preference in the settings for the user to select between automatic and manual start of the track recording with the navigation.

what do you think of it?

**Comment by Rob-from-VI** [comment]
Yes, this feature is much anticipated,  I currently have to use Maps Me for this purpose, which works quite well.

Very important to be able to start track recording from any point without navigation, as I would use it to record hiking trails that may or may not exist to be navigated.

Also, it would be VERY useful if the tracks could be exported in a format which is acceptable to open maps.   This means it has to include timestamps on each recorded point.   Not sure if that might be possible???

Some adjustable degree of smoothing of adjacent points might be useful as well, due to limitations of Phone GPS accuracy when hiking in forested areas.   (or averaging of several adjacent readings for each recorded point???)   Otherwise trail recordings can become very jagged, and this artificially inflates the length of the hike.

Not sure why one would want automatic start???   Some sort of reminder that recording is ongoing might be useful, as I have often recorded my hike, and then forgot to turn it off and save it, so I recorded my drive home as well!!!!

I would be happy to do beta testing of this feature if it is helpful.

**Comment by sandy99405** [comment]
Thank you @Rob-from-VI for the add-ons and explanations. It would be very beneficial if you were the beta tester of this feature :)


```````````
Very important to be able to start track recording from any point without navigation, as I would use it to record hiking trails that may or may not exist to be navigated.
````````````
Yes that is a good point. Recording can help when there is no navigation.


```````````
Also, it would be VERY useful if the tracks could be exported in a format which is acceptable to open maps. This means it has to include timestamps on each recorded point. Not sure if that might be possible???
```````````

Anyway, the records will be saved in GPX format, and also let's ensure that each point along the recorded track includes a timestamp indicating when it was recorded.


```````````
Some adjustable degree of smoothing of adjacent points might be useful as well, due to limitations of Phone GPS accuracy when hiking in forested areas. (or averaging of several adjacent readings for each recorded point???) Otherwise trail recordings can become very jagged, and this artificially inflates the length of the hike.
```````````

I am thinking of an algorithm called **Moving Average** which can be used for accurately smoothing the track adjacent points.


````````````
Not sure why one would want automatic start??? Some sort of reminder that recording is ongoing might be useful, as I have often recorded my hike, and then forgot to turn it off and save it, so I recorded my drive home as well!!!!
````````````

Reminding that recording is ongoing is a cool thing. So what I'm proposing is: Let's keep the automatic Record On by default but give the user an option to choose between the Automatic and the Manual Recordings in the general settings because if the user doesn't want to record the track the Record list might get clumsy and filled with unnecessary records.


Please let me know what you think about these.

**Comment by biodranik** [comment]
It may be easier to start by restoring the previously worked functionality of the "Recent track recorder" that displayed the recent track in the last 24 hours. This feature is already implemented in the C++ core and works on iOS. It requires a proper foreground service implementation to work. It can be a very good start/pre-requisite to a fully working active track recorder described in this issue. See https://github.com/organicmaps/organicmaps/issues/7477 for details.

**Comment by sandy99405** [comment]
Yes @biodranik I have already gone through this issue https://github.com/organicmaps/organicmaps/commit/2bfb7ffc0b1a060cf72d929ffd5d18779553d1aa before.
Correct me if I'm wrong, but the UI seems to be in the general settings which might not be conveniently visible to the users.

``````````
prefs_main.xml

<androidx.preference.PreferenceScreen
      android:key="@string/pref_track_screen"
      android:title="@string/pref_track_record_title"
      android:persistent="false"
      android:order="8">
      <SwitchPreferenceCompat
        android:key="@string/pref_track_record_time"
        android:title="@string/pref_track_record_title"
        android:order="1"/>
      <ListPreference
        android:key="@string/pref_track_record"
        android:title="@string/pref_track_record_title"
        android:entries="@array/track_length"
        android:entryValues="@array/track_length_values"
        android:order="2"/>
      <Preference
        android:selectable="false"
        android:persistent="false"
        android:enabled="false"
        android:summary="@string/recent_track_help_text"
        android:order="3"/>
    </androidx.preference.PreferenceScreen>
``````````

So, do you say it's good idea to proceed with this ?

**Comment by biodranik** [comment]
As I mentioned, let's start with restoring the feature that _was working_. Then it can be improved.

**Comment by sandy99405** [comment]
HI @biodranik  this is going to be a Test Driven Development as it helps in proper maintenance and bugs can be avoided.
Is this fine with you? Do you have anything else to suggest to/guide me about this Track Recorder feature?

**Comment by biodranik** [comment]
Did you read the whole thread? Check how @shankarpriyank researched it above. Evaluate the complexity of the feature. Split it into different subtasks. Find good enough UX solutions. Plan how to implement and test all corner cases.

**Comment by sandy99405** [comment]
Yes @biodranik , I have gone through the thread. I have been preparing a proposal for this feature. I will finish it asap. I just need a medium to send it to you so that you can review it. Please tell me where I can send it.

**Comment by biodranik** [comment]
Use hello at organicmaps dot app

Students who carefully fix and test issues on Github have more chances to be selected for GSoC.

**Comment by sandy99405** [comment]
Thank you @biodranik for providing the email. I have sent the proposal for this feature. Kindly check it and let me know if any improvements required.

**Comment by biodranik** [comment]
The old code can be checked as an example, if necessary. There is no goal to "revive old outdated code". The goal is to "release a useful feature for users".

**Comment by sonic-onkaringale** [comment]
Hello @biodranik  and @rtsisyk , I have previously worked on this type of problem in one of my applications published on the Play Store, and I am aware of the limitations of the location foreground service. I apologize for being a bit late for GSoC. However, I have submitted the proposal. Additionally, I have sent my proposal and my CV to hello@organicmaps.app.

**Comment by BikeRun** [comment]
There is a great iOS app „**Open GPX Tracker**“ that **can record gpx tracks and export them**, but has no offline maps. Its source code is available on github. **Maybe you can join to not reinvent the wheel.** Having the track recording feature from „Open GPX Tracker“ in organic maps would be great!

**Comment by biodranik** [comment]
The c++ core already has a working on iOS recent track recorder. Enabling it for Android should be easy, if a foreground service is implemented. 

Each GPS point is now taking 65 bytes in the recorded track file. Did anyone already check the c++ code? What exactly is stored there and in which format? What is missing and not stored?

**Comment by Rob-from-VI** [comment]
Downloaded latest beta which allowed me to record a "recernt track", which was simply a walk around my house, which then displayed over the map.

For now there seems to be no way to adjust the frequency of recorded points.

I was also unable to find a way to save the track after I stopped the recording.

I wonder about the need for predetermined length of recorded track??

**Comment by Rob-from-VI** [comment]
Greetings!!
      
      As a frequent user of this feature in other apps, I have been
      reading with some despair all of the numerous comments and
      discussions on this subject.  Some items that keep coming up have
      me puzzled:   I hope the following comments might be useful, to
      get something out there for us to use before we die of old age....
      :-)
      
      1)   Auto start recording when navigating.   Why on earth
      would anyone want to record a track that can be
      established by selecting a start and destination point (route) for
      navigating existing roads or paths???  A future option to save and
      re-open a multi-point navigation route could be useful, but that
      is a different issue altogether.  I don't know if anyone has such
      a thing.
      
      2)  Autostart recording in any fashion whatsoever.  I
      would definitely not want the program to record anything unless I
      have selected to record my track, and with something visible on
      screen to show that recording is taking place.  Main
      menu is just fine for "start recording" - eventually an on-sceen
      interface could be added, but definitely not essential.
      
      2) The endless discussions over the name of the thing being
      recorded.   As a user, I really don't care very much what you
      decide to call it, as long as it hints at the possibility that I
      will be able to save and share the route I am about to take. 
      "Track Recorder" is good - go with it!!!   
      
      3)  Any relationship to existing trails or roads:   My
      recorded tracks often but definitely not always follow sections of
      already mapped paths, but also sections not yet existing on the
      map.   The existence or absence of mapped paths or roads should
      initially have no effect on Recording.  Eventually future you
      could add a user select-able option to have the track "follow"
      existing routes, might be useful in some cases.
      
      4)  Time limit based recording:  I would much prefer not
      having to worry about a preset time limit, unless there is danger
      of running out of resources to save the track.  Initially just the
      above-mentioned indication that recording is taking place would be
      sufficient.  Possibly a notification that I have been recording
      for an extended period of time, do i wish to continue, but still
      continue recording until explicitly terminated by user. 
      Eventually, in future maybe Recording should also terminate and be
      autosaved whenever the app is closed.
      
      5)  Two different versions ie recent path (display only) and
        track recorder (saved tracks)  It seems totally
      counterproductive to have two options when one can serve both
      purposes, and all development and refinement effort can focus on
      just one set of tracking and interface.   Start recording (whether
      by menu choice, onscreen button, or whatever, will automatically
      display the "recent track" as it does in IOS now.   I think IOS
      users will easily learn to turn it on when they wish to see their
      recent path.  The only new bit is to be able to save it at
      completion or possibly eventually at increments if there is demand
      for that.
      
      6) Various options to stop recording and save the track: 
      Seems to me it is very important to have an option the edit the
      track name before saving, and ideally have a specific default list
      where these are saved so they can be found, managed, and
      re-displayed or exported.
      
      7) Eventually, in future, it would be very nice to have ability to
      trim existing tracks front and back, in case one has forgotten to
      stop recording at the end of the hike.  Another future feature
      that would be very useful would be to be able to move recorded
      points to correct or smooth a poorly recorded track, and possibly
      some options to set the frequency of points limited by minimum
      time between points and minimum distance, such that it does not
      record 20 points while i have stopped at a viewpoint. 
      
      So in conclusion, I am really looking forward to a very simple
      "stop recording, save track" interface to appear in beta or
      otherwise, so we can start using it.
      
      Then we can start agitating for the most desired future
      improvements. :-)
      
      Rob
      
      
      
      
      
      
      
      Alexander Borsuk wrote on 2024-05-12 12:52 p.m.:
    
    
      
      
      The c++ core already has a working on iOS recent
        track recorder. Enabling it for Android should be easy, if a
        foreground service is implemented.
      Each GPS point is now taking 65 bytes in the
        recorded track file. Did anyone already check the c++ code? What
        exactly is stored there and in which format? What is missing and
        not stored?
      —
        Reply to this email directly, view it on GitHub, or unsubscribe.
        You are receiving this because you were mentioned.Message
          ID: ***@***.***>
      [
{
***@***.***": "http://schema.org",
***@***.***": "EmailMessage",
"potentialAction": {
***@***.***": "ViewAction",
"target": "https://github.com/organicmaps/organicmaps/issues/2147#issuecomment-2106357122",
"url": "https://github.com/organicmaps/organicmaps/issues/2147#issuecomment-2106357122",
"name": "View Issue"
},
"description": "View this Issue on GitHub",
"publisher": {
***@***.***": "Organization",
"name": "GitHub",
"url": "https://github.com"
}
}
]
